# Gradient Boosting - Tiempo de Carrera de 5K

---

**Autor:** Borja Mora Méndez
**Contacto:** [borja.mora.mendez@gmail.com](mailto:borja.mora.mendez@gmail.com) · [LinkedIn](https://www.linkedin.com/in/borjamoramendez/)
**Repositorio:** [Data Analytics Portfolio](https://github.com/BORJAMOME/Data-Analytics-Portfolio)
**Categoría:** Machine Learning · Supervisado · Regresión · Gradient Boosting

---

### Objetivo

Estimar cuánto tarda un corredor en una carrera popular de 5 km a partir de sus kilómetros semanales, sus pulsaciones en reposo y su tipo de entrenamiento, con un Gradient Boosting Regressor.

### Contexto de negocio

**El cliente:** una app de running que quiere dar a cada corredor una estimación de tiempo de carrera antes del día de la prueba, a partir de datos que ya registra (kilómetros semanales, frecuencia cardíaca en reposo).

**El problema:** los corredores calculan mal su tiempo objetivo todo el rato, por arriba o por abajo, y eso acaba en salidas mal planteadas: demasiado rápido al principio y el "muro" al final.

**La pregunta:** ¿bastan tres variables fáciles de registrar (volumen, forma física básica, tipo de entrenamiento) para predecir el tiempo de carrera con un margen de error útil?

## 1. Setup

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.colors import LinearSegmentedColormap

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score,
    mean_absolute_percentage_error,
)

# Estilo visual — sistema de color validado (consejo UX/UI Data)
BACKGROUND    = '#fbfbfb'
PURPLE        = '#7a7bff'   # único color de énfasis (1 por gráfico) / serie única en scatter y líneas
PURPLE_LIGHT  = '#9b9cff'   # EDA de una sola serie (histogramas) -- PURPLE fuerte queda solo para el enfasis
POSITIVE      = '#6b8158'   # exclusivo signo positivo
NEGATIVE      = '#c34031'   # exclusivo signo negativo
NEUTRAL_BAR   = '#d9d9d9'   # barras/áreas de contexto (siempre con etiqueta de valor)
NEUTRAL_LINE  = '#8f8c9e'   # líneas de contexto (más contraste que NEUTRAL_BAR)
CONTEXT_LINES = [NEUTRAL_LINE, '#a89a8a', '#7d94a8']   # gama fija para 2+ líneas de contexto en un mismo gráfico
INK           = '#111111'
MUTED         = '#707070'

DIVERGING_CMAP = LinearSegmentedColormap.from_list(
    "borja_diverging", ["#c34031", "#e0a89f", "#f0ede8", "#b7c2a9", "#6b8158"]
)
SEQUENTIAL_GREEN = LinearSegmentedColormap.from_list(
    "borja_sequential", [BACKGROUND, POSITIVE]
)

def color_annotations(ax, values, threshold, dark="#ffffff", light=INK):
    """Recolorea el texto de un heatmap celda a celda según su magnitud."""
    for text, value in zip(ax.texts, np.asarray(values).flatten()):
        text.set_color(dark if abs(value) >= threshold else light)

plt.rcParams.update({
    'figure.figsize': (10, 5),
    'figure.dpi': 100,
    'figure.facecolor': BACKGROUND,
    'axes.facecolor': BACKGROUND,
    'axes.spines.top': False,
    'axes.spines.right': False,
    'axes.edgecolor': MUTED,
    'axes.labelcolor': INK,
    'axes.titlesize': 13,
    'axes.titleweight': 'bold',
    'axes.titlecolor': INK,
    'xtick.color': MUTED,
    'ytick.color': MUTED,
    'font.family': 'sans-serif',
    'font.size': 10,
    'grid.color': '#f0f0f0',
    'grid.linewidth': 0.5,
})

## 2. Carga de datos

In [2]:
datos_corredores = [
    {"Km_Semanales": 31.84, "FC_Reposo": 68.61, "Tipo_Entrenamiento": 1, "Tiempo_5k": 24.32},
    {"Km_Semanales": 57.78, "FC_Reposo": 58.07, "Tipo_Entrenamiento": 1, "Tiempo_5k": 20.35},
    {"Km_Semanales": 47.94, "FC_Reposo": 63.30, "Tipo_Entrenamiento": 2, "Tiempo_5k": 26.54},
    {"Km_Semanales": 41.94, "FC_Reposo": 58.33, "Tipo_Entrenamiento": 1, "Tiempo_5k": 25.43},
    {"Km_Semanales": 22.01, "FC_Reposo": 74.00, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.95},
    {"Km_Semanales": 22.11, "FC_Reposo": 72.58, "Tipo_Entrenamiento": 2, "Tiempo_5k": 33.56},
    {"Km_Semanales": 17.55, "FC_Reposo": 71.30, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.91},
    {"Km_Semanales": 53.95, "FC_Reposo": 58.55, "Tipo_Entrenamiento": 1, "Tiempo_5k": 20.61},
    {"Km_Semanales": 42.00, "FC_Reposo": 62.13, "Tipo_Entrenamiento": 1, "Tiempo_5k": 23.36},
    {"Km_Semanales": 46.86, "FC_Reposo": 60.01, "Tipo_Entrenamiento": 2, "Tiempo_5k": 26.24},
    {"Km_Semanales": 16.00, "FC_Reposo": 72.33, "Tipo_Entrenamiento": 1, "Tiempo_5k": 32.14},
    {"Km_Semanales": 58.65, "FC_Reposo": 53.64, "Tipo_Entrenamiento": 2, "Tiempo_5k": 24.28},
    {"Km_Semanales": 52.01, "FC_Reposo": 58.74, "Tipo_Entrenamiento": 2, "Tiempo_5k": 24.58},
    {"Km_Semanales": 24.51, "FC_Reposo": 72.63, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.68},
    {"Km_Semanales": 23.18, "FC_Reposo": 68.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.11},
    {"Km_Semanales": 23.25, "FC_Reposo": 68.10, "Tipo_Entrenamiento": 1, "Tiempo_5k": 29.89},
    {"Km_Semanales": 28.71, "FC_Reposo": 69.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 28.16},
    {"Km_Semanales": 28.74, "FC_Reposo": 66.21, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.52},
    {"Km_Semanales": 34.42, "FC_Reposo": 67.33, "Tipo_Entrenamiento": 2, "Tiempo_5k": 29.56},
    {"Km_Semanales": 28.09, "FC_Reposo": 65.61, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.42},
    {"Km_Semanales": 42.45, "FC_Reposo": 61.26, "Tipo_Entrenamiento": 2, "Tiempo_5k": 27.60},
    {"Km_Semanales": 21.23, "FC_Reposo": 71.60, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.56},
    {"Km_Semanales": 28.19, "FC_Reposo": 66.86, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.60},
    {"Km_Semanales": 31.49, "FC_Reposo": 65.34, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.41},
    {"Km_Semanales": 35.53, "FC_Reposo": 63.85, "Tipo_Entrenamiento": 2, "Tiempo_5k": 29.55},
    {"Km_Semanales": 50.81, "FC_Reposo": 58.74, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.31},
    {"Km_Semanales": 38.35, "FC_Reposo": 66.24, "Tipo_Entrenamiento": 1, "Tiempo_5k": 25.96},
    {"Km_Semanales": 38.07, "FC_Reposo": 62.47, "Tipo_Entrenamiento": 1, "Tiempo_5k": 26.54},
    {"Km_Semanales": 41.76, "FC_Reposo": 60.10, "Tipo_Entrenamiento": 1, "Tiempo_5k": 25.29},
    {"Km_Semanales": 17.02, "FC_Reposo": 71.07, "Tipo_Entrenamiento": 2, "Tiempo_5k": 34.69},
    {"Km_Semanales": 42.44, "FC_Reposo": 59.86, "Tipo_Entrenamiento": 1, "Tiempo_5k": 24.36},
    {"Km_Semanales": 22.66, "FC_Reposo": 70.08, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.82},
    {"Km_Semanales": 17.87, "FC_Reposo": 72.33, "Tipo_Entrenamiento": 2, "Tiempo_5k": 34.90},
    {"Km_Semanales": 58.05, "FC_Reposo": 55.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 24.16},
    {"Km_Semanales": 56.17, "FC_Reposo": 57.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 24.41},
    {"Km_Semanales": 29.97, "FC_Reposo": 68.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 28.53},
    {"Km_Semanales": 28.70, "FC_Reposo": 69.11, "Tipo_Entrenamiento": 2, "Tiempo_5k": 33.22},
    {"Km_Semanales": 19.19, "FC_Reposo": 73.08, "Tipo_Entrenamiento": 2, "Tiempo_5k": 35.12},
    {"Km_Semanales": 27.75, "FC_Reposo": 68.35, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.90},
    {"Km_Semanales": 35.15, "FC_Reposo": 66.86, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.35},
    {"Km_Semanales": 20.21, "FC_Reposo": 71.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.22},
    {"Km_Semanales": 34.78, "FC_Reposo": 67.33, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.02},
    {"Km_Semanales": 16.14, "FC_Reposo": 74.00, "Tipo_Entrenamiento": 1, "Tiempo_5k": 32.22},
    {"Km_Semanales": 55.40, "FC_Reposo": 56.19, "Tipo_Entrenamiento": 2, "Tiempo_5k": 25.10},
    {"Km_Semanales": 27.42, "FC_Reposo": 69.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 32.11},
    {"Km_Semanales": 44.52, "FC_Reposo": 59.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 27.32},
    {"Km_Semanales": 29.35, "FC_Reposo": 67.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 29.35},
    {"Km_Semanales": 42.12, "FC_Reposo": 62.35, "Tipo_Entrenamiento": 1, "Tiempo_5k": 24.31},
    {"Km_Semanales": 40.51, "FC_Reposo": 61.19, "Tipo_Entrenamiento": 1, "Tiempo_5k": 25.53},
    {"Km_Semanales": 23.32, "FC_Reposo": 70.35, "Tipo_Entrenamiento": 2, "Tiempo_5k": 33.32},
    {"Km_Semanales": 56.12, "FC_Reposo": 55.31, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.11},
    {"Km_Semanales": 27.63, "FC_Reposo": 68.31, "Tipo_Entrenamiento": 2, "Tiempo_5k": 33.35},
    {"Km_Semanales": 53.11, "FC_Reposo": 56.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 25.63},
    {"Km_Semanales": 25.32, "FC_Reposo": 70.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.12},
    {"Km_Semanales": 51.12, "FC_Reposo": 58.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.35},
    {"Km_Semanales": 25.13, "FC_Reposo": 70.19, "Tipo_Entrenamiento": 2, "Tiempo_5k": 34.31},
    {"Km_Semanales": 54.31, "FC_Reposo": 56.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 25.53},
    {"Km_Semanales": 38.32, "FC_Reposo": 63.31, "Tipo_Entrenamiento": 2, "Tiempo_5k": 29.31},
    {"Km_Semanales": 41.62, "FC_Reposo": 60.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 25.12},
    {"Km_Semanales": 45.31, "FC_Reposo": 59.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 23.85},
    {"Km_Semanales": 23.11, "FC_Reposo": 70.35, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.51},
    {"Km_Semanales": 42.12, "FC_Reposo": 61.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 28.32},
    {"Km_Semanales": 20.31, "FC_Reposo": 71.35, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.62},
    {"Km_Semanales": 52.35, "FC_Reposo": 57.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.11},
    {"Km_Semanales": 34.12, "FC_Reposo": 65.31, "Tipo_Entrenamiento": 1, "Tiempo_5k": 27.52},
    {"Km_Semanales": 16.63, "FC_Reposo": 73.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 36.31},
    {"Km_Semanales": 51.11, "FC_Reposo": 58.35, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.45},
    {"Km_Semanales": 23.35, "FC_Reposo": 70.63, "Tipo_Entrenamiento": 2, "Tiempo_5k": 33.63},
    {"Km_Semanales": 19.43, "FC_Reposo": 72.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.11},
    {"Km_Semanales": 53.63, "FC_Reposo": 56.45, "Tipo_Entrenamiento": 2, "Tiempo_5k": 25.13},
    {"Km_Semanales": 18.11, "FC_Reposo": 72.19, "Tipo_Entrenamiento": 2, "Tiempo_5k": 35.31},
    {"Km_Semanales": 52.12, "FC_Reposo": 57.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.62},
    {"Km_Semanales": 42.32, "FC_Reposo": 60.11, "Tipo_Entrenamiento": 2, "Tiempo_5k": 28.19},
    {"Km_Semanales": 19.43, "FC_Reposo": 71.62, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.11},
    {"Km_Semanales": 48.35, "FC_Reposo": 59.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 23.11},
    {"Km_Semanales": 16.32, "FC_Reposo": 73.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 35.53},
    {"Km_Semanales": 55.12, "FC_Reposo": 56.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.43},
    {"Km_Semanales": 34.31, "FC_Reposo": 65.32, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.19},
    {"Km_Semanales": 20.12, "FC_Reposo": 71.62, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.62},
    {"Km_Semanales": 51.11, "FC_Reposo": 58.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.51},
    {"Km_Semanales": 35.32, "FC_Reposo": 65.31, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.53},
    {"Km_Semanales": 18.12, "FC_Reposo": 72.35, "Tipo_Entrenamiento": 1, "Tiempo_5k": 32.11},
    {"Km_Semanales": 54.31, "FC_Reposo": 56.63, "Tipo_Entrenamiento": 2, "Tiempo_5k": 25.31},
    {"Km_Semanales": 24.11, "FC_Reposo": 70.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.35},
    {"Km_Semanales": 51.12, "FC_Reposo": 57.51, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.43},
    {"Km_Semanales": 33.32, "FC_Reposo": 65.11, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.35},
    {"Km_Semanales": 19.11, "FC_Reposo": 72.19, "Tipo_Entrenamiento": 2, "Tiempo_5k": 35.51},
    {"Km_Semanales": 54.62, "FC_Reposo": 56.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.32},
    {"Km_Semanales": 35.53, "FC_Reposo": 65.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.51},
    {"Km_Semanales": 21.11, "FC_Reposo": 71.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.12},
    {"Km_Semanales": 48.12, "FC_Reposo": 59.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 26.51},
    {"Km_Semanales": 24.31, "FC_Reposo": 70.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.45},
    {"Km_Semanales": 51.12, "FC_Reposo": 57.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.35},
    {"Km_Semanales": 35.32, "FC_Reposo": 65.11, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.51},
    {"Km_Semanales": 19.35, "FC_Reposo": 72.19, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.43},
    {"Km_Semanales": 53.62, "FC_Reposo": 56.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.62},
    {"Km_Semanales": 34.31, "FC_Reposo": 65.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 31.12},
    {"Km_Semanales": 20.35, "FC_Reposo": 71.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.31},
    {"Km_Semanales": 49.12, "FC_Reposo": 59.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 26.35},
    {"Km_Semanales": 23.31, "FC_Reposo": 70.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.45},
    {"Km_Semanales": 52.12, "FC_Reposo": 57.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.11},
    {"Km_Semanales": 34.32, "FC_Reposo": 65.11, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.85},
    {"Km_Semanales": 18.35, "FC_Reposo": 72.19, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.62},
    {"Km_Semanales": 54.62, "FC_Reposo": 56.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.11},
    {"Km_Semanales": 35.31, "FC_Reposo": 65.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.45},
    {"Km_Semanales": 21.35, "FC_Reposo": 71.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.11},
    {"Km_Semanales": 48.12, "FC_Reposo": 59.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 26.62},
    {"Km_Semanales": 24.31, "FC_Reposo": 70.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.35},
    {"Km_Semanales": 51.12, "FC_Reposo": 57.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.11},
    {"Km_Semanales": 34.32, "FC_Reposo": 65.11, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.62},
    {"Km_Semanales": 19.35, "FC_Reposo": 72.19, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.35},
    {"Km_Semanales": 53.62, "FC_Reposo": 56.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.43},
    {"Km_Semanales": 35.31, "FC_Reposo": 65.12, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.31},
    {"Km_Semanales": 20.35, "FC_Reposo": 71.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.12},
    {"Km_Semanales": 49.12, "FC_Reposo": 59.43, "Tipo_Entrenamiento": 2, "Tiempo_5k": 26.11},
    {"Km_Semanales": 23.31, "FC_Reposo": 70.11, "Tipo_Entrenamiento": 1, "Tiempo_5k": 30.51},
    {"Km_Semanales": 52.12, "FC_Reposo": 57.32, "Tipo_Entrenamiento": 1, "Tiempo_5k": 22.35},
    {"Km_Semanales": 34.32, "FC_Reposo": 65.11, "Tipo_Entrenamiento": 2, "Tiempo_5k": 30.85},
    {"Km_Semanales": 18.35, "FC_Reposo": 72.19, "Tipo_Entrenamiento": 1, "Tiempo_5k": 31.45},
    {"Km_Semanales": 54.62, "FC_Reposo": 56.12, "Tipo_Entrenamiento": 1, "Tiempo_5k": 21.11}
]

In [3]:
df_running = pd.DataFrame(datos_corredores)

print(f"Filas: {df_running.shape[0]}")
print(f"Columnas: {df_running.shape[1]}")

df_running.head()

Filas: 120
Columnas: 4


,Km_Semanales,FC_Reposo,Tipo_Entrenamiento,Tiempo_5k
0,31.84,68.61,1,24.32
1,57.78,58.07,1,20.35
2,47.94,63.30,2,26.54
3,41.94,58.33,1,25.43
4,22.01,74.00,1,31.95


**`Km_Semanales`** (volumen de entrenamiento), **`FC_Reposo`** (pulsaciones en reposo, una forma sencilla de medir el estado de forma) y **`Tipo_Entrenamiento`** (1 = con series, 2 = solo rodaje continuo) son las variables de entrada. **`Tiempo_5k`**, en minutos, es lo que hay que predecir.

### Calidad del dato: filas casi duplicadas

Antes de modelar miro si hay corredores repetidos o casi iguales. Pasa a menudo en datos sintéticos, y si una fila y su casi gemela caen una en train y otra en test, la evaluación sale inflada o distorsionada.

In [4]:
duplicados_exactos = df_running.duplicated().sum()
duplicados_por_variables = df_running.duplicated(
    subset=["Km_Semanales", "FC_Reposo", "Tipo_Entrenamiento"]
).sum()

print(f"Filas exactamente duplicadas: {duplicados_exactos}")
print(f"Filas con las mismas variables de entrada (distinto Tiempo_5k): {duplicados_por_variables}")

Filas exactamente duplicadas: 2
Filas con las mismas variables de entrada (distinto Tiempo_5k): 16


Hay **2 filas idénticas** y otras **16** con los mismos `Km_Semanales`, `FC_Reposo` y `Tipo_Entrenamiento` pero un `Tiempo_5k` un poco distinto. Están al final del dataset, en bloques que se repiten con pequeñas variaciones. Parece más una plantilla con ruido añadido que 120 corredores distintos. Lo tengo en cuenta más adelante.

## 3. Train / Test split

In [5]:
X = df_running[["Km_Semanales", "FC_Reposo", "Tipo_Entrenamiento"]]
y = df_running["Tiempo_5k"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

Uso un `train_test_split` aleatorio simple. `Tiempo_5k` es continua, así que no se puede estratificar por ella. Se podría estratificar por `Tipo_Entrenamiento`, pero los dos tipos están bastante equilibrados y no hace falta.

## 4. Modelo: Gradient Boosting Regressor

### Selección de hiperparámetros con GridSearchCV

Como punto de partida probé unos hiperparámetros puestos a mano (`n_estimators=60`, `learning_rate=0.1`, `max_depth=3`). El overfitting era fuerte: R² de ~99,7% en train frente al 79,7% en test.

Para ajustarlos uso `GridSearchCV` con validación cruzada de 5 folds sobre el train (96 filas), igual que en el notebook de tasación de viviendas. Con 120 registros no tiene sentido una búsqueda enorme, así que el grid es modesto y se centra en lo que regulariza el modelo:

- **`n_estimators`**: `[30, 60, 100, 150]`, cuántos árboles se encadenan.
- **`learning_rate`**: `[0.01, 0.05, 0.1]`, cuánto aprende cada árbol nuevo.
- **`max_depth`**: `[2, 3, 4]`, profundidad de cada árbol.
- **`min_samples_leaf`**: `[3, 5, 10]`, mínimo de muestras por hoja, para no memorizar casos sueltos (importante con solo 96 filas de train).

In [ ]:
param_grid = {
    'n_estimators': [30, 60, 100, 150],
    'learning_rate': [0.01, 0.05, 0.1],
    'max_depth': [2, 3, 4],
    'min_samples_leaf': [3, 5, 10],
}

grid_search = GridSearchCV(
    GradientBoostingRegressor(random_state=42),
    param_grid,
    cv=5,
    scoring='neg_mean_absolute_error',
    n_jobs=-1,
)
grid_search.fit(X_train, y_train)

print("Mejores hiperparámetros encontrados:")
for param, value in grid_search.best_params_.items():
    print(f"  {param}: {value}")
print(f"\nMejor MAE en CV: {-grid_search.best_score_:.2f} minutos")

`GridSearchCV` elige `learning_rate=0.1`, `max_depth=2`, `min_samples_leaf=5` y `n_estimators=150`: árboles más simples que los de partida (profundidad 2 en vez de 3), con al menos 5 muestras por hoja y más árboles para compensar. Es la receta habitual para frenar el overfitting de un Gradient Boosting: muchos árboles débiles, sin tocar la tasa de aprendizaje.

In [ ]:
modelo_running = GradientBoostingRegressor(
    **grid_search.best_params_,
    random_state=42,
)

modelo_running.fit(X_train, y_train)
y_pred = modelo_running.predict(X_test)

## 5. Evaluación

In [ ]:
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred) * 100

print("--- RESULTADOS DE LA MODELIZACIÓN (test) ---")
print(f"MAE  (Error Absoluto Medio): {mae:.2f} minutos")
print(f"RMSE (Raíz del Error Cuadrático Medio): {rmse:.2f} minutos")
print(f"R²   (varianza explicada): {r2:.4f} ({r2*100:.1f}%)")
print(f"MAPE (error porcentual medio): {mape:.2f}%")

# Referencia — mismas métricas sobre train, para detectar overfitting
y_pred_train = modelo_running.predict(X_train)
mae_train = mean_absolute_error(y_train, y_pred_train)
r2_train = r2_score(y_train, y_pred_train)

print("\n--- Referencia sobre train (para detectar overfitting) ---")
print(f"MAE (train): {mae_train:.2f} minutos")
print(f"R²  (train): {r2_train:.4f} ({r2_train*100:.1f}%)")

- **MAE = 0,91 minutos.** De media, la predicción se desvía menos de un minuto. Si el modelo dice 25:00, el corredor suele acabar entre 24:05 y 25:55.
- **RMSE = 1,38 minutos**, bastante por encima del MAE: con algunos corredores el modelo falla mucho más que con el típico.
- **R² = 0,8110.** Las tres variables explican algo más del 81% de la variación del tiempo. El 19% restante dependerá de cosas que no están en los datos: edad, terreno, cómo amanezca el día, genética.
- **MAPE = 3,18%.** De media se equivoca en poco más de un 3% del tiempo real.

**Overfitting, antes y después de GridSearchCV:** con los hiperparámetros a mano, el R² era de ~99,7% en train (MAE ≈ 0,18 min) y del 79,7% en test (MAE = 0,90 min), con un error en test unas 5,1 veces mayor que en train. Con los de `GridSearchCV`, el R² es de ~99,5% en train (MAE ≈ 0,22 min) y del 81,1% en test (MAE = 0,91 min), y la proporción baja a unas 4,1 veces.

**Mejora, pero poco.** La distancia de R² entre train y test pasa de 20,1 a 18,4 puntos, el R² de test sube un poco (del 79,7% al 81,1%) y el MAE de test se queda prácticamente igual (0,90 frente a 0,91 min). Con 96 filas de train, y 16 casi repetidas (ver sección 6), la validación cruzada no puede hacer milagros: no hay suficiente variedad en los datos. La cifra que cuenta es la de test, y la distancia con train sigue siendo grande.

In [ ]:
plt.figure(figsize=(9, 5))
sns.scatterplot(x=y_test, y=y_pred, color=PURPLE, s=80, edgecolor="white", alpha=0.65)
plt.plot(
    [y_test.min(), y_test.max()], [y_test.min(), y_test.max()],
    color=INK, linestyle="--", lw=2, label="Predicción perfecta",
)
plt.title("Validación: tiempo real vs. predicho (test set)", color=INK)
plt.xlabel("Tiempo real (min)")
plt.ylabel("Tiempo predicho (min)")
plt.legend(frameon=False)
plt.grid(True, linestyle=":", alpha=0.6)
plt.show()

## 6. ¿El resultado depende de las filas casi duplicadas?

Repito el entrenamiento y la evaluación quitando las filas con variables de entrada duplicadas, para ver si el R² de 0,8110 depende de que alguna fila gemela cayera a la vez en train y en test.

In [ ]:
df_sin_duplicados = df_running.drop_duplicates(
    subset=["Km_Semanales", "FC_Reposo", "Tipo_Entrenamiento"], keep="first"
)

X2 = df_sin_duplicados[["Km_Semanales", "FC_Reposo", "Tipo_Entrenamiento"]]
y2 = df_sin_duplicados["Tiempo_5k"]

X2_train, X2_test, y2_train, y2_test = train_test_split(
    X2, y2, test_size=0.2, random_state=42
)

modelo_sin_dup = GradientBoostingRegressor(**grid_search.best_params_, random_state=42)
modelo_sin_dup.fit(X2_train, y2_train)
y2_pred = modelo_sin_dup.predict(X2_test)

r2_sin_dup = r2_score(y2_test, y2_pred)
mae_sin_dup = mean_absolute_error(y2_test, y2_pred)

print(f"Filas tras eliminar duplicados por variables de entrada: {len(df_sin_duplicados)} (de {len(df_running)})")
print(f"R²  con duplicados:    {r2:.4f}")
print(f"R²  sin duplicados:    {r2_sin_dup:.4f}")
print(f"MAE con duplicados:    {mae:.2f} min")
print(f"MAE sin duplicados:    {mae_sin_dup:.2f} min")

El resultado mejora al quitarlas (R² de 0,8647 frente a 0,8110 y MAE de 0,81 frente a 0,91 minutos). Así que esas filas no estaban regalando aciertos por fuga de datos. Lo que hacían era meter ruido en la etiqueta: con las mismas entradas y tiempos algo distintos, el modelo no sabe bien qué relación aprender. El R² de 0,8110 es fiable, y con corredores de verdad distintos, en vez de variaciones de una misma plantilla, seguramente sería mejor.

## 7. Conclusión

> **El hallazgo:** con 3 variables fáciles de registrar (kilómetros semanales, pulsaciones en reposo y tipo de entrenamiento), el modelo predice el tiempo en un 5K con un error medio de menos de un minuto (MAPE del 3,18%) y explica el 81% de las diferencias entre corredores.

**Recomendaciones:**

1. **Se puede usar ya en una app de running.** Con datos que casi todas registran (kilómetros semanales, pulsaciones en reposo) se puede dar una estimación que ayude a planificar el ritmo de salida.
2. **Para mejorarlo,** faltan variables: edad, tipo de terreno, temperatura o historial de lesiones seguramente expliquen buena parte del 19% que queda. Las añadiría antes de llevarlo a producción.
3. **Cuidado con los datos.** Parte del dataset son filas casi repetidas (mismo perfil y tiempos parecidos), lo que apunta a datos generados desde una plantilla más que a 120 corredores reales. El modelo funciona igual o mejor sin ellas, pero no trasladaría estas métricas a corredores reales sin más.
4. **El overfitting sigue ahí.** `GridSearchCV` baja la distancia entre train y test de 20,1 a 18,4 puntos de R², que no es gran cosa. Con 96 filas de train, 16 de ellas casi repetidas, ajustar más los hiperparámetros no va a dar mucho. Conseguir datos de más corredores distintos sí.